src/runtime/gpu_context_common.h

```diff
@@ -0,0 +1,192 @@
+#include "printer.h"
+#include "scoped_mutex_lock.h"
+
+namespace Halide {
+namespace Internal {
+
+template<typename ContextT, typename ModuleStateT>
+struct GPUCompilationCache {
+    struct CachedCompilation {
+        ContextT context;
+        ModuleStateT module_state;
+        uint32_t kernel_id;
+    };
+
+    halide_mutex mutex;
+
+    static constexpr float kLoadFactor = .5f;
+    static constexpr int kInitialTableBits = 7;
+    int log2_compilations_size{0};  // number of bits in index into compilations table.
+    CachedCompilation *compilations{nullptr};
+    int count{0};
+
+    static constexpr uint32_t kInvalidId{0};
+    static constexpr uint32_t kDeletedId{1};
+
+    uint32_t unique_id{2};  // zero is an invalid id
+
+    static ALWAYS_INLINE uintptr_t kernel_hash(ContextT context, uint32_t id, uint32_t bits) {
+        uintptr_t addr = (uintptr_t)context + id;
+        // Fibonacci hashing. The golden ratio is 1.9E3779B97F4A7C15F39...
+        // in hexadecimal.
+        if (sizeof(uintptr_t) >= 8) {
+            return (addr * (uintptr_t)0x9E3779B97F4A7C15) >> (64 - bits);
+        } else {
+            return (addr * (uintptr_t)0x9E3779B9) >> (32 - bits);
+        }
+    }
+
+    bool insert(ContextT context, uint32_t id, ModuleStateT module_state) {
+        if (log2_compilations_size == 0) {
+            if (!resize_table(kInitialTableBits)) {
+                return false;
+            }
+        }
+        if ((count + 1) > (1 << log2_compilations_size) * kLoadFactor) {
+            if (!resize_table(log2_compilations_size + 1)) {
+                return false;
+            }
+        }
+        count += 1;
+        uintptr_t index = kernel_hash(context, id, log2_compilations_size);
+        for (int i = 0; i < (1 << log2_compilations_size); i++) {
+            uintptr_t effective_index = (index + i) & ((1 << log2_compilations_size) - 1);
+            if (compilations[effective_index].kernel_id <= kDeletedId) {
+                compilations[effective_index].context = context;
+                compilations[effective_index].module_state = module_state;
+                compilations[effective_index].kernel_id = id;
+                return true;
+            }
+        }
+        // This is a logic error that should never occur. It means the table is
+        // full, but it should have been resized.
+        halide_assert(nullptr, false);
+        return false;
+    }
+
+    bool find_internal(ContextT context, uint32_t id, ModuleStateT *&module_state) {
```

Comment:
nit: Do we have a MUST_USE_RESULT macro available in our runtime code? If so, marking `find_internal` and `lookup` that way seems prudent.